[New plugin] PixArt TP

hughsie opened the pull request:
### Plugin Overview
This plugin adds support for PixArt Touchpad devices using custom HID-I2C
communication.

### Update Protocol Summary
- Reset to bootloader mode (0xaa/0xcc)
- Internal flash erase and 256-byte page programming
- Firmware CRC and parameter CRC verification after flashing
- Reset back to application mode after update

### CI and Test Coverage
A full emulated device has been uploaded to LVFS, enabling fwupd CI to run
`emulate` update tests for this plugin.

Physical hardware CI is not available because HID-I2C touchpad devices cannot be
connected to GitHub CI infrastructure.

### Checklist
- [x] Fill out README.md with update protocol
- [x] Fill out README.md with any custom quirks and flags
- [x] Implement FuFirmware->write() and include at least one fuzzer testcase
- [x] CI run of the plugin for at least one target (LVFS emulate device)
- [x] Document targets CI isn't currently run and the reasons
- [x] Update CODEOWNERS

PixartHarris committed:
feat: add pixart touchpad plugin

PixartHarris committed:
feat: update Readme.md

hughsie committed:
fix codeowners

PixartHarris committed:
fix: fix the pre-commit error.

PixartHarris committed:
fix : fix pxi-tp.json  pretty format json format error.

PixartHarris committed:
fix: fix No newline at end of file error for pxi-tp-setup.json.

PixartHarris committed:
fix: add missing newline at end of JSON file to satisfy formatting/linting rules

hughsie committed:
fix debug prefix -- already set with G_LOG_DOMAIN

hughsie commented on plugins/pixart-tp/fu-pxi-tp-common.c:
```diff
@@ -0,0 +1,77 @@
+}
+
+gboolean
+fu_pxi_tp_common_propagate_with_log(GError **error, GError **error_local, const char *ctx)
```
@PixartHarris again, please use g_propagate_error() directly -- not this.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-firmware.c:
```diff
@@ -0,0 +1,953 @@
+
+// =============================================================
+// file: fu-pxi-tp-firmware.c (pure C, offset-based parsing; no packed structs)
+// =============================================================
```
@PixartHarris no // comments please -- and this kind of comment isn't really useful either

hughsie commented on plugins/pixart-tp/fu-pxi-tp-firmware.c:
```diff
@@ -0,0 +1,953 @@
+		}
+		if (dstcap > 0)
+			s->external_file_name[want] = '\0';
+		_fu_trim_ascii_nul(s->external_file_name, dstcap);
```
does fu_strsafe() work here?

hughsie committed:
fix debug prefix -- already set with G_LOG_DOMAIN

hughsie commented on plugins/pixart-tp/fu-pxi-tp-register.c:
```diff
@@ -0,0 +1,144 @@
+	guint8 buf[4] = {REPORT_ID_SINGLE, addr, bank, val};
+	return fu_pxi_tp_common_send_feature(self, buf, sizeof(buf), error);
+}
+gboolean
```
I think running ./contrib/reformat-code.py on this subdirectory might be a good idea

hughsie commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,913 @@
+	part_id += out_val << 8;
+
+	switch (part_id) {
+	case 0x0274:
```
random number!

hughsie commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,913 @@
+	fu_progress_step_done(progress);
+
+	prog_verify = fu_progress_get_child(progress);
+	(void)flags;
```
you don't need to do this -- ignored parameters are 100% fine.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-firmware.c:
```diff
@@ -0,0 +1,953 @@
+}
+
+static const char *
+fu_pxi_tp_firmware_update_type_to_str(guint8 t)
```
I do think you want to put PXI_TP_UPDATE_TYPE_GENERAL and friends in the .rs file -- then you get this kind of code generated at build time for free -- e.g. read https://fwupd.github.io/libfwupdplugin/tutorial.html#the-rustgen-helper and it will make a lot of the other code make sense too.

PixartHarris replied:
This has been addressed — the PXI_TP_UPDATE_TYPE_* enums have been moved into the .rs file and are now generated by rustgen.
Please review and let me know if this can be resolved on your side.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-firmware.c:
```diff
@@ -0,0 +1,953 @@
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), NULL);
+	return self->sections;
+}
```
I think you're reinvented `fu_firmware_get_images()` :) -- I think each section should be a "child" -- e.g. use 

    FuFirmware *img = fu_firmware_new()
    if (!fu_firmware_parse(img, the_section_bytes, error))
        return FALSE;
    fu_firmware_add_image(firmware, img);

when you're parsing the sections...

hughsie commented on plugins/pixart-tp/fu-pxi-tp-tf-communication.c:
```diff
@@ -0,0 +1,707 @@
+	gsize len = 3; /* expected payload length = 3 bytes (version) */
+
+	for (gsize i = 0; i < FAILED_RETRY_TIMES; i++) {
+		g_autoptr(GError) local_error = NULL;
```
@PixartHarris this should use fu_device_retry() instead of this loop.

hughsie commented on plugins/pixart-tp/pixart-tp.quirk:
```diff
@@ -0,0 +1,17 @@
+Plugin = pixart_tp
+GType = FuPxiTpDevice
+HidVersionReg = bank=0x00;addr=0xb2
+SramSelect = 0x0f
```
this needs to have a PxiTp prefix -- e.g. make it `PxiTpSramSelect` -- it also has to be declared in `fu_pxi_tp_plugin_constructed()` with `fu_context_add_quirk_key()` too

hughsie commented on plugins/pixart-tp/fu-pxi-tp-tf-communication.c:
```diff
@@ -0,0 +1,707 @@
+				target_ver[1],
+				target_ver[2]);
+			return TRUE;
+		}
```
> to manage two separate firmware versions

The way other plugins do it is have a "child device" -- e.g. for the synaptics prometheus device there's a "main device" and a "config device" -- I think that would be more natural here too -- i.e. you have a Touchpad device and a Haptic device -- with the Haptic writes going through the Touchpad device as a "proxy". This way they can both have different version numbers, and you don't need to worry about upreving versions of the touchpad to get the haptic device updated.

PixartHarris replied:
Hi @hughsie 
I have a question about how child devices should determine whether a firmware update is actually needed.

For the parent TP device, this is straightforward because the metainfo.xml contains a <release version="..."> entry, and fwupd can compare the installed version against the metadata version to decide if an update is required.

However, for the child device, I can read the firmware version from the IC during setup(), but I cannot find any mechanism to define the file version for the child device’s firmware image in the metainfo file.
The XML only specifies the parent firmware version, for example:
```xml
<releases>
  <release version="0xff02" date="2025-10-27" urgency="medium">
    <description>
      <p>Testing used firmware.</p>
    </description>
    <checksum type="sha256" target="content" filename="firmware.bin">
      3646be4720845dd9654d77577a4a37814b7447facd724d9000235d2dfd02678c
    </checksum>
  </release>
</releases>
```
Since no separate <release> exists for the child device, how should the child device determine whether an update is necessary?
Is fwupd’s current behavior that child devices always update whenever the parent package is updated, regardless of the child device’s own version?

Or is there a recommended way for the plugin to provide a "file version" for the child image so that fwupd can skip child updates when the versions match?

Thanks!

hughsie replied:
> By giving each metainfo its own GUID

Exactly that. But rather than a -tf suffix, can you do something like `&COMPONENT_tf` i.e. use `fu_device_add_instance_str(FU_DEVICE(self), "COMPONENT", "tf")` in the TP `_init()` -- then you can use `fu_device_build_instance_id()` to build the instance ID.

> fwupd shows a shared progress bar rather than separate progress bars

Yup, completely by design (and a lot of work!) -- to the user they're only updating "one thing" so one transition from 0..100 makes sense to them.

PixartHarris committed:
refactor: simplify error handling and register access
- Drop custom fu_pxi_tp_common_* helpers and register macros
- Use fu_pxi_tp_register*() functions with proper GError propagation
- Replace fu_pxi_tp_common_fail() with g_set_error[_literal]() and prefixes

PixartHarris committed:
refactor: clean magic numbers and use standard CRC8 implementation

- remove nocheck:magic-inlines suppression
- replace magic numbers with named constants
- remove custom CRC8_TABLE and use fu_crc8(FU_CRC_KIND_B8_STANDARD)

PixartHarris committed:
fix: avoid re-reading pixart-tp version in attach

- Skip fu_pxi_tp_device_setup() on normal attach when not in bootloader
  mode, as the version is already set in ->setup().
- When returning from bootloader, keep the version refresh but treat it
  as best-effort and log with g_debug instead of g_warning so
  G_DEBUG=fatal-criticals does not abort self-tests.
- Runtime behaviour on real hardware is unchanged; this only reduces log
  noise and fixes the fu-self-test SIGTRAP from "failed to refresh
  version (already attached)".

PixartHarris committed:
refactor: pixart-tp: replace HidVersionReg with explicit quirk keys

The previous HidVersionReg mini-DSL required string parsing and made the
quirk handling unnecessarily complex. Replace it with three explicit
quirk keys:

 • PxiTpHidVersionBank
 • PxiTpHidVersionAddr
 • PxiTpSramSelect

This follows the fwupd quirk convention of one key per value and avoids
manual parsing. The new keys are registered in
fu_pxi_tp_plugin_constructed() so that they can be consumed by the core
fwupd quirk loader.

PixartHarris committed:
refactor: use Rustgen structs for TF RMI headers

Generate TF pass-through header structs from fu-pxi-tp.rs and use them
in the TF communication helpers instead of hand-written offsets.

The new Rustgen file defines three packed structs:

 * FuStructPxiTfWriteSimpleCmd
 * FuStructPxiTfWritePacketCmd
 * FuStructPxiTfReadCmd

These map directly to the TF write/read header layouts and provide
typed setters for all fields. The C implementation in
fu-pxi-tp-tf-communication.c now builds TF commands by:

 * allocating the corresponding FuStructPxiTf*Cmd
 * filling fields with the generated setters (endian-safe)
 * copying the header into the feature report buffer
 * appending payload, CRC and tail as before

To avoid a name clash with the existing firmware struct helper, the
old fu-pxi-tp-struct.[ch] files have been renamed to
fu-pxi-tp-fw-struct.[ch], and all includes and meson.build sources
have been updated accordingly. The Rustgen output continues to use the
fu-pxi-tp-struct.[ch] name.

Firmware update logic and TF protocol behavior are unchanged; only the
header construction has been refactored for better readability and
maintainability, addressing the review feedback about “magic offsets”
and missing endian-safe struct definitions.

PixartHarris committed:
refactor: Export firmware metadata as hex only

Use fu_xmlb_builder_insert_kx() directly for all numeric fields in the
FWHD exporter instead of exporting both hex and decimal variants.
Drop fu_pxi_tp_firmware_kx_and_dec() helper and the *_dec metadata keys.

hughsie committed:
less header includes

hughsie commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1321 @@
+	PXI_TP_SECTOR_SIZE = 4096,
+	PXI_TP_PAGE_SIZE = 256,
+	PXI_TP_PAGES_COUNT_PER_SECTOR = 16,
+};
```
i don't think this is an enum -- they're just three # define values. enums are supposed to form a "set" if you see what I mean.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1321 @@
+};
+
+/* ---- CRC control for firmware ---- */
+enum {
```
if you give this a name then you can use it as a type for fu_pxi_tp_register_write() -- if you have two guint8 variables it's easy to swap them around, but if you have `FuPxiTpCrc` and `FuPxiBankId` then the compiler will warn you if they're wrong.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-register.c:
```diff
@@ -0,0 +1,249 @@
+		return FALSE;
+	}
+
+	error_local = NULL;
```
this is a bit of an antipattern -- you don't need error_local here at all. Just do:

	if (!fu_pxi_tp_common_send_feature(self, cmd, sizeof(cmd), error)) {
		g_prefix_error(error,
			    "register read command failed: bank=0x%02x addr=0x%02x: ",
			    bank,
			    addr);
		return FALSE;
	}

hughsie commented on plugins/pixart-tp/fu-pxi-tp-register.c:
```diff
@@ -0,0 +1,249 @@
+			    1, /* src offset: skip report id */
+			    copy_len,
+			    &error_local);
+	if (!ok) {
```
you don't need ok here -- just put the fu_memcpy_safe in the `if ()`

hughsie commented on plugins/pixart-tp/fu-pxi-tp.rs:
```diff
@@ -0,0 +1,64 @@
+#[derive(New)]
+#[repr(C, packed)]
+struct FuStructPxiTfWritePacketCmd {
+    report_id: u8,
```
if this is always REPORT_ID__PASS_THROUGH then you can define it here -- add `Defaults` to the list in `#[derive()]` and then use `report_id: u8 == 0xCC` -- this also means the `fu_struct_pxi_tf_write_packet_cmd_set_report_id()` command won't be generated anymore, because `fu_struct_pxi_tf_write_packet_cmd_new()` will set the value by default.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-tf-communication.c:
```diff
@@ -0,0 +1,842 @@
+	fu_struct_pxi_tf_read_cmd_set_report_id(st_read, REPORT_ID__PASS_THROUGH);
+	fu_struct_pxi_tf_read_cmd_set_preamble(st_read, TF_FRAME_PREAMBLE);
+	fu_struct_pxi_tf_read_cmd_set_slave_addr(st_read, SLAVE_ADDRESS);
+	fu_struct_pxi_tf_read_cmd_set_func(st_read, TF_FUNC_READ_WITH_LEN);
```
i think these 4 are not needed if you set the defaults in the .rs file

hughsie commented on plugins/pixart-tp/fu-pxi-tp-firmware.c:
```diff
@@ -0,0 +1,940 @@
+				    &self->file_ver,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
```
I think using `#[derive(ParseStream)]` and defining `FuPxiTpFirmwareHdr` in the .rs file would make this much nicer indeed!

PixartHarris committed:
refactor: replace non-enum values with defines and consts for clarity

PixartHarris committed:
refactor: move version formatting to version_raw + convert_version()

Use fu_firmware_set_version_raw() and a convert_version() vfunc instead
of formatting the hex string directly in parse().

PixartHarris committed:
refactor: use fu_firmware_get_version_raw() instead of custom getter

PixartHarris committed:
refactor: use fu_device_retry_full() for TF read helpers

- replace manual retry loops in fu_pxi_tp_tf_communication_read_firmware_version()
  with fu_device_retry_full()
- replace manual retry loops in fu_pxi_tp_tf_communication_read_download_status()
  with fu_device_retry_full()
- add small context structs and static retry callbacks for version and
  download-status reads
- keep higher-level TF update attempt logic unchanged

PixartHarris committed:
refactor: use FuChunkArray for TF firmware writes

- Replace manual offset/length math with fu_chunk_array_new_from_bytes in TF RMI write path
- Treat empty firmware payload as no-op and re-enable touch before returning
- Ensure touch is re-enabled on chunk-array creation/index failures and packet write failures
- Keep 1-based packet index for protocol while iterating 0-based chunk array

PixartHarris committed:
pixart-tp: move TP enums to Rust and use rustgen macros

Moved all TP reset/flash/CRC enums from C into fu-pxi-tp.rs and switched
fu-pxi-tp-device.c to use the generated FU_PXI_TP_* macros. Updated
reset/flash/CRC logic accordingly. TF enums unchanged.

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1321 @@
+
+typedef enum {
+	PXI_TP_RESET_MODE_APPLICATION = 0,
+	PXI_TP_RESET_MODE_BOOTLOADER = 1,
```
Hi,
I want to double-check my understanding regarding which enums should be moved into the Rust .rs file (so they get exported as FU_PXI_TP_* macros), and which ones should remain in C and be given a proper typedef enum type.
For example, currently in C I have:
```c
typedef enum {
    FU_PXI_TP_RESET_MODE_APPLICATION = 0,
    FU_PXI_TP_RESET_MODE_BOOTLOADER  = 1,
} FuPxiTpResetMode;
```
And in Rust this would become:
```rust
enum FuPxiTpResetMode {
    Application,
    Bootloader,
}
```
Should this particular enum (ResetMode) be moved into the .rs file, or should it stay in C and be used as a typed enum via typedef enum?

I’m a bit confused about which enums belong in Rust and which should stay in C.
From your earlier comments, I understand that having strongly-typed enums in C helps prevent mixing parameters such as bank IDs, CRC modes, etc. But I’m not sure which ones fwupd prefers to be Rust-generated and which ones should remain as C types.

Could you please clarify which categories of enums you expect to be placed in .rs (and exported via rustgen), and which enums should remain in C with a named typedef enum?

Thanks!

hughsie replied:
If you put them in the .rs file they actually get created as C typedefs too -- so there's no need to typedef them in C also. e.g. `plugins/hughski-colorhug/fu-hughski-colorhug.rs` has:

	#[derive(ToString)]
	enum FuHughskiColorhugError {
	    None,
	    UnknownCmd,
	    ...,
	    SelfTestEeprom,
	}

...which gets generated as `venv/build/plugins/hughski-colorhug/libfu_plugin_hughski_colorhug.a.p/fu-hughski-colorhug-struct.h` with:

	typedef enum {
	    FU_HUGHSKI_COLORHUG_ERROR_NONE,
	    FU_HUGHSKI_COLORHUG_ERROR_UNKNOWN_CMD,
	    ...,
	    FU_HUGHSKI_COLORHUG_ERROR_SELF_TEST_EEPROM,
	}  FuHughskiColorhugError;
	#define FU_HUGHSKI_COLORHUG_ERROR_LAST 36
	const gchar *fu_hughski_colorhug_error_to_string(FuHughskiColorhugError val) G_GNUC_CONST;

... so then you can hash `include "fu-hughski-colorhug-struct.h"` and get the C enum and any of the auto-generated code you defined with derive.

PixartHarris committed:
refactor: replace error_local with direct error propagation in fu-pxi-tp-register.c

- Simplify error handling and use g_prefix_error()/g_prefix_error_literal() where appropriate.

PixartHarris committed:
refactor: simplify error handling in fu-pxi-tp-register.c

PixartHarris committed:
refactor: modify TF communication constants and rustgen structs

- Replace old TF_* macros/enums with FU_PXI_TF_* constants and update call sites
- Move TF command/frame/touch/upgrade constants into Rust enums
- Use rustgen defaults for report_id/preamble/slave_addr/func to shrink C code

PixartHarris committed:
refactor: rename slave_addr to target_addr in TF structs

PixartHarris committed:
refactor: remove dead code and unnecessary NULL checks

PixartHarris committed:
refactor: build TF write frames directly in GByteArray

Remove temporary stack buffer and append payload/CRC/tail directly to the
rustgen GByteArray before sending. Simplifies code and matches review
suggestion.

PixartHarris committed:
refactor: rely on cleanup() for TF rollback

PixartHarris committed:
refactor: fix C90 mixed declarations in TF update path

PixartHarris committed:
refactor : pad TF frames to 64 bytes and add upgrade-mode exit helper

- Pad TF write-simple and write-packet frames with zero bytes to ensure
  full 64-byte feature-report frames before sending to the device.
- Add fu_pxi_tp_tf_communication_exit_upgrade_mode() as a shared helper
  for leaving TF upgrade/bootloader mode.
- Use the new helper in both the firmware update flow and device cleanup
  to ensure APP-version reads do not return the bootloader version.

PixartHarris committed:
refactor: type banks and registers via rustgen

- Move bank IDs and register offsets into fu-pxi-tp.rs and export as
  FU_PXI_TP_SYSTEM_BANK_*/FU_PXI_TP_USER_BANK_* and FU_PXI_TP_REG_* macros
- Change register helpers to use FuPxiTpSystemBank/FuPxiTpUserBank instead
  of raw guint8
- Update reset/flash/CRC/TF proxy callers to use the new typed bank and
  register macros

PixartHarris committed:
refactor: fix C90 forbids mixed declarations and code [-Wdeclaration-after-statement] warning

PixartHarris committed:
refactor: parse FWHD header via rustgen structs

- Use FuStructPxiTpFirmwareHdr/SectionHdr instead of manual offset parsing
- Move FWHD layout constants into fu-pxi-tp.rs and clean up C90 warnings

PixartHarris committed:
refactor: remove unused feature helpers and inline HIDRAW calls

The fu_pxi_tp_common_{send,get}_feature helpers were redundant as all
callers already set buf[0] to the REPORT_ID before issuing the ioctl.
Replace them with direct fu_hidraw_device_{set,get}_feature() calls.

PixartHarris committed:
pixart-tp: use rustgen enums and expose section metadata

- Replace anonymous enums with #defines for fixed FWHD sizes and limits
- Export TF HID/RMI constants as #[repr(u8)] enums and use them in TF
  write/read structs instead of raw u8 literals
- Rename FWHD section reserved field to shared in rustgen, use the
  generated getter to copy shared bytes into FuPxiTpSection::reserved
- Call exit_upgrade_mode after TF download so we read the application
  FW version rather than the bootloader version

hughsie committed:
fix up the README file a bit

hughsie committed:
style fixes

PixartHarris committed:
refactor: replace manual flash/CRC polling loops with fu_device_retry_full()
Changes include:
 - Add flash_execute_wait_cb() and use fu_device_retry_full() in
   fu_pxi_tp_device_flash_execute()
 - Add flash_write_enable_wait_cb() and replace manual WEL polling loop
 - Add flash_wait_busy_wait_cb() and replace busy-wait loop during flash ops
 - Add crc_firmware_wait_cb() and replace CRC busy polling loop
 - Remove duplicated sleep/poll logic from all affected functions
 - Use g_prefix_error_literal() to add context on final failure

PixartHarris committed:
refactor: modify CRC helpers and write_firmware flow

- Change fu_pxi_tp_device_crc_firmware() and fu_pxi_tp_device_crc_parameter()
  to return gboolean and use a guint32* out parameter so CRC=0 is valid
- Add fu_pxi_tp_device_crc_parameter_wait_cb() and use fu_device_retry_full()
  for parameter CRC busy polling
- Factor out fu_pxi_tp_device_verify_crc() to handle firmware/parameter CRC
  comparison and progress reporting
- Simplify fu_pxi_tp_device_write_firmware() by delegating CRC verification to
  the new helper and renaming locals for better readability

PixartHarris committed:
refactor: simplify page write logic and always initialize buffer to 0xFF

The previous implementation used two branches depending on whether the
remaining payload exactly matched the page size. This made the code harder
to follow and duplicated the memcpy logic.

This change unifies the flow:
- Always pre-fill the page buffer with 0xFF (matching flash erase state)
- Copy only the valid payload bytes using fu_memcpy_safe()
- Remove conditional branches and reduce code complexity

Behavior is unchanged, but the function is now clearer and easier to
maintain.

PixartHarris committed:
refactor: implement prepare_firmware() and remove legacy wrapper

Implement the missing fu_pxi_tp_device_prepare_firmware() vfunc so the
core can correctly create and parse FuPxiTpFirmware from the input
stream. This avoids the need to manually reload and parse the firmware
during write() and eliminates the old workaround helper
fu_pxi_tp_device_wrap_or_parse_ctn().

This brings the plugin in line with the expected fwupd firmware
lifecycle and prevents double-parsing of the CAB payload.

PixartHarris committed:
refactor: use raw version and add convert_version callback

Use fu_device_set_version_raw() instead of manually formatting the
version string, allowing fwupd to store the canonical numeric version.
Add a convert_version() implementation to format the device version as
0xABCD and expose it through the device_class->convert_version vfunc.

Also update class_init to register the new callback and clean up the
klass_device variable naming.

PixartHarris committed:
refactor: read version using fu_memread_uint16()

Replace manual uint16 assembly (lo/hi shifting) with a small
two-byte buffer and fu_memread_uint16() to decode the LE version
field. This improves readability, handles endianness correctly, and
matches the fwupd coding style.

Also updates setup() to store the result via fu_device_set_version_raw().

PixartHarris committed:
Merge branch 'pixart/pixart_tp' of https://github.com/fwupd/fwupd into pixart/pixart_tp

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1306 @@
+				    "flash write enable failure");
+		g_debug("flash write enable failure");
+		return FALSE;
+	}
```
Addressed in commit [cfd4ac6c4d9279fa3bfda69981fcc0a12f5c5cce](https://github.com/fwupd/fwupd/pull/9624/commits/cfd4ac6c4d9279fa3bfda69981fcc0a12f5c5cce)

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1306 @@
+		return FALSE;
+
+	crc_value = fu_pxi_tp_device_crc_firmware(self, error);
+	if (error != NULL && *error != NULL)
```
Addressed in commit [b2404bc245dfb123cfa3fb4ac39a72a0e6b29a3f](https://github.com/fwupd/fwupd/pull/9624/commits/b2404bc245dfb123cfa3fb4ac39a72a0e6b29a3f)

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1306 @@
+		(guint)ver_u16);
+
+	ver_str = g_strdup_printf("0x%04x", ver_u16);
+	fu_device_set_version(device, ver_str);
```
Addressed in commit [8cdc0f29ad73215f57919cd3ba4c8ab84ce994fe](https://github.com/fwupd/fwupd/pull/9624/commits/8cdc0f29ad73215f57919cd3ba4c8ab84ce994fe)

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1306 @@
+{
+	FuPxiTpDevice *self = FU_PXI_TP_DEVICE(device);
+	guint8 lo = 0;
+	guint8 hi = 0;
```
Addressed in commit [cf111ef12329ef1ee05ad1d63f8a5911a7278123](https://github.com/fwupd/fwupd/pull/9624/commits/cf111ef12329ef1ee05ad1d63f8a5911a7278123)

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -0,0 +1,1306 @@
+				    error))
+			return FALSE;
+	} else {
+		memset(page_buf, 0xFF, sizeof(page_buf));
```
For now, I’m proceeding with this approach. If this behavior does not match what you intended, I can adjust it again in a follow-up commit.

Addressed in commit [a835920123bda2b125a924ee76945aa484495801](https://github.com/fwupd/fwupd/pull/9624/commits/a835920123bda2b125a924ee76945aa484495801).

hughsie replied:
Yup, that's fine -- you can also do `guint8 page_buf[PXI_TP_PAGE_SIZE] = {[0 ... 63] = 0xFF};` but that's kinda ugly. Even better would be to do `g_autoptr(GByteArray) page_buf = g_byte_array_new()` and then `fu_byte_array_set_size(page_buf, PXI_TP_PAGE_SIZE, 0xFF);`

PixartHarris replied:
I’ve updated the implementation to use g_byte_array_new() and fu_byte_array_set_size() instead.

Addressed in commit [fb7f8597cef15b23fd68f9dab8083b11f9ebe6f6](https://github.com/fwupd/fwupd/pull/9624/commits/fb7f8597cef15b23fd68f9dab8083b11f9ebe6f6)

PixartHarris commented:
Hi @hughsie 
To speed up development, I would like to add my colleague PixartMicky to this PR so that he can also push fixes directly to the branch pixart/pixart_tp.
Could you please grant him access to commit to this branch?
Thank you!

PixartMicky committed:
refactor: modify section writing and fix progress steps

Move section write loop into fu_pxi_tp_device_write_sections() and correct
progress handling to avoid over-consuming child steps.

PixartHarris committed:
refactor: fix various compiler warnings and improve page-buffer handling

- Mark CRC helpers as static to avoid missing-prototype warnings
- Switch write_page() to GByteArray and adjust memcpy_safe usage
- Fix unused-parameter warning by annotating `flags` in export()
- Resolve C90 mixed-declaration warning by moving g_autofree declaration
- Make write_sections() take const GPtrArray* for correctness

PixartHarris committed:
refactor: add detailed debug tracing for detach/attach/reload/cleanup

This patch adds additional g_debug() tracing around the bootloader and
engineer-mode transitions to make the update flow easier to diagnose.
The previous logs were insufficient to determine whether the device was
entering or exiting bootloader mode at the correct time, especially when
self-tests or emulated devices were used.

Changes include:
 • Log when setup() is called
 • Add structured debug messages in detach(), attach(), reload(), cleanup()
 • Trace entry/exit of bootloader mode and IS_BOOTLOADER flag state
 • Trace TF engineer-mode exit attempts
 • Register the reload() vfunc for post-update version refresh

hughsie commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -1054,6 +1054,51 @@ fu_pxi_tp_device_process_section(FuPxiTpDevice *self,
+
+	g_debug("Section len: %u", sections->len);
+
+	for (section_idx = 0; section_idx < sections->len; section_idx++) {
```
just `(guint i = 0...` please

hughsie commented:
> refactor: add detailed debug tracing for detach/attach/reload/cleanup

I don't like that commit -- at all. There is a huge amount of debugging in the core code and we don't want the same thing in each plugin.

PixartMicky committed:
refactor: use fu_progress_get_child() in fu_pxi_tp_device_write_sections()

Simplify progress handling by passing fu_progress_get_child(progress)
directly to fu_pxi_tp_device_write_sections(). This removes manual child
progress construction and keeps the function aligned with other callers
using the updated progress API.

PixartMicky commented on plugins/pixart-tp/fu-pxi-tp-device.c:
```diff
@@ -1054,6 +1054,51 @@ fu_pxi_tp_device_process_section(FuPxiTpDevice *self,
+{
+	guint section_idx = 0;
+	FuProgress *progress_write = NULL;
+	progress_write = fu_progress_get_child(progress);
```
Address in [88a1f5eb996ae7db821a9127f1800d46fa12278d](https://github.com/fwupd/fwupd/commit/88a1f5eb996ae7db821a9127f1800d46fa12278d)

superm1 commented on plugins/pixart-tp/pixart-tp.quirk:
```diff
@@ -0,0 +1,20 @@
+GType = FuPxiTpDevice
+PxiTpHidVersionBank = 0x00
+PxiTpHidVersionAddr = 0xb2
+PxiTpSramSelect = 0x0f
```
logically I would think sort by HWID

PixartHarris commented on plugins/pixart-tp/meson.build:
```diff
@@ -0,0 +1,60 @@
+  install_data(
+    ['tests/pxi-tp.builder.xml'],
+    install_dir: join_paths(installed_test_datadir, 'tests'),
+    install_tag: 'tests',   # ★ add tag for installed-test data
```
Thanks for pointing this out — the install_tag isn’t necessary here.
I’ve removed it in the updated patch.

superm1 commented on plugins/pixart-tp/fu-pxi-tp-plugin.c:
```diff
@@ -0,0 +1,55 @@
+{
+	FuPlugin *plugin = FU_PLUGIN(obj);
+	fu_plugin_set_name(plugin, "pixart_tp");
+}
```
Yeah; that's my expectation.

hughsie replied:
tbh, I'd like to rename it to match the others -- I've happy to do this for you when we're ready for merge. e.g. see https://github.com/fwupd/fwupd/pull/9643 -- at the moment getting the code right is more important than the prefix.

PixartHarris committed:
refactor: reduce verbose debug logging in attach/detach/reload/cleanup

Remove per-step g_debug() tracing from the plugin hotpath to avoid
duplicated and noisy logs compared to existing fwupd core diagnostics.

The updated logic:
 - only logs ignored failures (best-effort paths)
 - no logging on normal success paths

PixartMicky committed:
refactor: Implement child progress handling in fu_pxi_tp_tf_communication_write_firmware

PixartHarris committed:
fix: use installed setup.json for emulation in pxi-tp test

Replace the incorrect ZIP file reference with the installed
pxi-tp-setup.json using @enumeration_datadir@ so the self-test
emulates the device correctly. This aligns the test with the
expected fwupd emulation workflow and matches other built-in plugins.

PixartMicky committed:
Merge branch 'pixart/pixart_tp' of https://github.com/fwupd/fwupd into pixart/pixart_tp

PixartHarris committed:
fix: install pxi-tp-setup.json as enumeration data

Move pxi-tp-setup.json from device-tests to enumeration-data so the
installed-tests framework can load it correctly for device emulation.

PixartHarris committed:
refactor: sort pixart quirk entries by HWID

PixartHarris committed:
refactor: use sizeof(payload) - 1 for burst write size check

PixartHarris committed:
cleanup: remove manual plugin name assignment

fu_plugin_set_name() is no longer required because the default plugin
name now matches the directory name (pixart-tp). The separate
object_constructed override can therefore be removed.

PixartHarris committed:
refactor: move update-type enum to Rust + migrate C callers

Move TP update-type definitions into fu-pxi-tp.rs so rustgen exports the
canonical enum. Update all C code to use FU_PXI_TP_UPDATE_TYPE_* and drop
the old PXI_TP_UPDATE_TYPE_* + local string helper.

Also restore fu_plugin_set_name("pixart_tp") since relying on the
directory-derived name (“pixart-tp”) caused device matching to stop.

PixartHarris committed:
pixart-tp: clean up README and remove outdated sections

Remove duplicated or obsolete content from the plugin README:
 - Drop old protocol ID reference.
 - Fix quirk key name to `PxiTpSramSelect`.
 - Remove outdated note about /dev/bus/usb access.

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-fw-struct.h:
```diff
@@ -0,0 +1,59 @@
+	PXI_TP_UPDATE_TYPE_PARAM = 3u,	    /* parameter section update */
+	PXI_TP_UPDATE_TYPE_TF_FORCE = 16u   /* TF Force (via DLL) */
+} FuPxiTpUpdateType;
+
```
Addressed in commit [0d700eab828f172c27a4533a2605d6eb4f1ff2c8](https://github.com/fwupd/fwupd/pull/9624/commits/0d700eab828f172c27a4533a2605d6eb4f1ff2c8)

PixartHarris commented on plugins/pixart-tp/README.md:
```diff
@@ -0,0 +1,90 @@
+
+This plugin supports the following protocol ID:
+
+- `com.pixart.tp`
```
Addressed in commit [fac5821b6aa8773e3809ad540750bcf63cd810a0](https://github.com/fwupd/fwupd/pull/9624/commits/fac5821b6aa8773e3809ad540750bcf63cd810a0)

PixartHarris committed:
refactor: use inline guint i loop index instead of section_idx

PixartHarris committed:
refactor: remove unused firmware/register helper functions

These functions are no longer used after the new section-based
write logic and TF communication refactor. Removing them avoids
unused symbol failures in CI.

PixartHarris committed:
refactor: switch TF packet write retry to fu_device_retry_full()

Changes:
 - add FuPxiTpTfWritePacketCtx to carry write parameters
 - add fu_pxi_tp_tf_communication_write_packet_cb() callback
 - replace per-packet retry loop with fu_device_retry_full()
 - keep outer update-attempt loop unchanged

PixartHarris committed:
refactor: split TP firmware sections into child images

PixartHarris committed:
feat: update the emulated device file

PixartHarris committed:
refactor: cleanup unused helpers and switch tests to emulation-url

- Remove the unused fu_pxi_tp_firmware_get_slice_by_file() helper
- Drop stale Chinese comments
- Convert FuPxiTpSection to use G_DEFINE_TYPE for better CI compatibility
- Update installed-tests to use emulation-url instead of emulation-file

hughsie commented on plugins/pixart-tp/fu-pxi-tp-section.c:
```diff
@@ -0,0 +1,311 @@
+{
+	memset(self->reserved, 0, sizeof self->reserved);
+	memset(self->external_file_name, 0, sizeof self->external_file_name);
+	self->flags = 0;
```
you don't need any of this -- GObject memsets the entire struct to 0 by default.

PixartHarris replied:
Addressed in commit [70c1547b47693fdd799a9d19751404374cd7dbae](https://github.com/fwupd/fwupd/pull/9624/commits/70c1547b47693fdd799a9d19751404374cd7dbae)

hughsie commented on plugins/pixart-tp/fu-pxi-tp-section.c:
```diff
@@ -0,0 +1,311 @@
+
+	FuPxiTpUpdateType update_type;
+	guint8 update_info;
+	guint32 flags; /* FuPxiTpFirmwareFlag bitmask */
```
Maybe `FuPxiTpFirmwareFlags flags`? You can call it `XxxxxFlags` in the .rs file and the constants are still singular.

hughsie commented on plugins/pixart-tp/fu-pxi-tp-fw-struct.h:
```diff
@@ -0,0 +1,32 @@
+	FU_PXI_TP_FIRMWARE_FLAG_NONE = 0,
+	FU_PXI_TP_FIRMWARE_FLAG_VALID = 1 << 0,
+	FU_PXI_TP_FIRMWARE_FLAG_IS_EXTERNAL = 1 << 1,
+} FuPxiTpFirmwareFlag;
```
into the .rs file please

PixartHarris replied:
Addressed in commit [88b6cc7d7086adf846d722dbba040e184e0072d3](https://github.com/fwupd/fwupd/pull/9624/commits/88b6cc7d7086adf846d722dbba040e184e0072d3)

hughsie commented on plugins/pixart-tp/fu-pxi-tp-fw-struct.h:
```diff
@@ -0,0 +1,32 @@
+
+/* ---- magic string ------------------------------------------------------- */
+/* used by validate() for quick reject before doing full parse */
+extern const char PXI_TP_MAGIC[5]; /* "FWHD" + NUL */
```
move this into fu-pxi-tp-firmware.c please

hughsie commented on plugins/pixart-tp/fu-pxi-tp-fw-struct.c:
```diff
@@ -0,0 +1,12 @@
+#include "fu-pxi-tp-struct.h"
+
+/* put into .rodata; visible to all TUs */
+const char PXI_TP_MAGIC[5] = "FWHD";
```
I don't think this is needed here

PixartHarris committed:
feat(pixart-tp): move TF_FORCE handling to haptic child device

- Add FuPxiTpHapticDevice child and create it from TP probe() when quirk PxiTpHasTfChild=1
- Parse TF_FORCE as a dedicated firmware image (id=com.pixart.tf-force) and update it via the child device
- TP parent now skips TF_FORCE sections and only programs TP firmware/parameter sections

PixartHarris committed:
refactor: drop custom hexdump helper and use fu_byte_array_to_string

PixartHarris committed:
refactor: remove redundant initialization and drop unused common helper

PixartHarris committed:
refactor: store section reserved bytes as a GByteArray

Replace the fixed-size reserved[] buffer in FuPxiTpSection with a GByteArray* and initialize it to a fixed length using fu_byte_array_set_size(). This allows using fu_byte_array_to_string() directly during export without creating temporary byte arrays.

Also add a finalize handler to unref the GByteArray,.

PixartHarris committed:
refactor: use FuFirmware filename for external section files

PixartHarris committed:
refactor: move TP firmware flags to rustgen and use typed Flags

PixartHarris committed:
refactor: Remove the legacy fu-pxi-tp-fw-struct helpers and derive firmware layout
information directly from the rustgen-generated structures.

The FWHD magic and firmware-level constants are now localized to
fu-pxi-tp-firmware.c, while section layout sizes are derived from
FuStructPxiTpFirmwareSectionHdr via rustgen-provided N_ELEMENTS
constants instead of duplicated macros.

The section reserved accessor is also updated to return a GByteArray
with clear ownership semantics, replacing the raw pointer + length API.

PixartHarris committed:
pixart-tp: Refactor TF protocol to use autogen structs and constants

Move TF (Touch Firmware) protocol definitions, timing constants, and frame
limits from C macros to Rust enums.

Updates include:
- Define `FuPxiTfTiming`, `FuPxiTfLimit`, and `FuPxiTfRetry` in Rust.
- Add `FuStructPxiTfReplyHdr`, `FuStructPxiTfVersionPayload`, and
  `FuStructPxiTfDownloadStatusPayload` structs for safer parsing.
- Refactor `fu-pxi-tp-tf-communication.c` to use the generated constants
  and struct helpers.
- Replace stack-allocated VLA-like padding with `g_byte_array_set_size`.
- Add boundary checks for ROM header validation.
- Remove explicit version comparison logic in favor of MCU download status checks.

PixartHarris committed:
refactor: Use standard parsing vfuncs for firmware sections

Refactor the firmware parsing logic to correctly use the FuFirmware
infrastructure:

1. Implement the `parse` virtual function in `FuPxiTpSection`. This allows
   the section to handle its own stream slicing using `fu_partial_input_stream_new`
   based on the offsets parsed from the descriptor.
2. Update `FuPxiTpFirmware` to call `fu_firmware_parse_stream()` instead of
   the custom `fu_pxi_tp_section_attach_payload_stream()` helper.
3. Simplify the validation logic in the parent firmware parser to fail
   immediately if a required section (FW or PARAM) is marked as invalid,
   rather than accumulating flags.
4. Stop reading the entire firmware blob into memory at the start; only
   read the header fixed bytes.

This fixes an issue where the firmware payload was not being correctly
set for write operations.

PixartHarris committed:
refactor: zero-pad TF feature reports instead of resizing buffer

The TF RMI feature report must be fully zero-padded to
FU_PXI_TF_FRAME_SIZE_FEATURE_REPORT_LEN before sending.

Previously this code used g_byte_array_set_size() to extend the
buffer, but that does not guarantee the newly-added bytes are
zeroed. On real hardware this usually works by chance, but on
emulated devices the uninitialized tail bytes cause ioctl
mismatches and test failures.

Fix this by explicitly appending a zero-filled buffer for the
remaining bytes, ensuring deterministic feature report contents.

This change fixes failures seen in emulated-device tests and
aligns the behavior with the actual device protocol expectations.

Also update the pxi-tp emulation setup and test JSON to reflect
the corrected IOCTL sequences and firmware artifacts.

PixartHarris committed:
refactor: remove unused local variables

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-section.c:
```diff
@@ -0,0 +1,311 @@
+	self->flags = flags;
+}
+
+/* ---------------------- public API ---------------------------- */
```
Removed in the latest commit.

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-section.c:
```diff
@@ -0,0 +1,311 @@
+fu_pxi_tp_section_class_init(FuPxiTpSectionClass *klass)
+{
+	FuFirmwareClass *firmware_class = FU_FIRMWARE_CLASS(klass);
+	firmware_class->export = fu_pxi_tp_section_export;
```
Addressed in commit [a1b15d98c5cb18cf573617a9bee31efb512d9abe](https://github.com/fwupd/fwupd/pull/9624/commits/a1b15d98c5cb18cf573617a9bee31efb512d9abe)

PixartHarris commented on plugins/pixart-tp/fu-pxi-tp-section.c:
```diff
@@ -0,0 +1,311 @@
+	guint32 section_length;
+	guint32 section_crc;
+
+	guint8 reserved[PXI_TP_S_RESERVED_LEN];
```
Addressed in commit [707a576e74d199b0a4e267ea644b0403fda79f2a](https://github.com/fwupd/fwupd/pull/9624/commits/707a576e74d199b0a4e267ea644b0403fda79f2a)
